File: backend/pb_migrations/1733842800_initial_schema.js
/// <reference path="../pb_data/types.d.ts" />
migrate((db) => {
  // Create announcements collection
  const announcements = new Collection({
    id: 'announcements_collection',
    name: 'announcements',
    type: 'base',
    system: false,
    schema: [
      {
        name: 'title',
        type: 'text',
        required: true,
        options: {
          min: 1,
          max: 200
        }
      },
      {
        name: 'slug',
        type: 'text',
        required: true,
        options: {
          min: 1,
          max: 200,
          pattern: '^[a-z0-9]+(?:-[a-z0-9]+)*$'
        }
      },
      {
        name: 'content',
        type: 'editor',
        required: true,
        options: {
          convertUrls: true
        }
      },
      {
        name: 'published_date',
        type: 'date',
        required: true
      },
      {
        name: 'featured',
        type: 'bool',
        required: false
      }
    ],
    indexes: [
      'CREATE UNIQUE INDEX idx_unique_slug ON announcements (slug)'
    ],
    listRule: '',  // Public read
    viewRule: '',  // Public read
    createRule: null,  // Admin only
    updateRule: null,  // Admin only
    deleteRule: null   // Admin only
  });

  db.collections.save(announcements);

  // Create documents collection
  const documents = new Collection({
    id: 'documents_collection',
    name: 'documents',
    type: 'base',
    system: false,
    schema: [
      {
        name: 'title',
        type: 'text',
        required: true,
        options: {
          min: 1,
          max: 200
        }
      },
      {
        name: 'category',
        type: 'select',
        required: true,
        options: {
          maxSelect: 1,
          values: [
            'Meeting Minutes',
            'Bylaws',
            'Financial Reports',
            'Architectural Guidelines'
          ]
        }
      },
      {
        name: 'year',
        type: 'number',
        required: true,
        options: {
          min: 2000,
          max: 2100
        }
      },
      {
        name: 'file',
        type: 'file',
        required: true,
        options: {
          maxSelect: 1,
          maxSize: 10485760,  // 10MB
          mimeTypes: ['application/pdf']
        }
      },
      {
        name: 'description',
        type: 'text',
        required: false,
        options: {
          max: 500
        }
      }
    ],
    indexes: [
      'CREATE INDEX idx_category ON documents (category)',
      'CREATE INDEX idx_year ON documents (year)'
    ],
    listRule: '',  // Public read
    viewRule: '',  // Public read
    createRule: null,  // Admin only
    updateRule: null,  // Admin only
    deleteRule: null   // Admin only
  });

  db.collections.save(documents);

  return true;
}, (db) => {
  // Rollback
  const announcements = db.findCollectionByNameOrId('announcements_collection');
  if (announcements) {
    db.deleteCollection(announcements);
  }

  const documents = db.findCollectionByNameOrId('documents_collection');
  if (documents) {
    db.deleteCollection(documents);
  }

  return true;
});

File: backend/migrations/1733842800_initial_schema.go
package migrations

import (
	"log"
	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/daos"
	m "github.com/pocketbase/pocketbase/migrations"
	"github.com/pocketbase/pocketbase/models"
	"github.com/pocketbase/pocketbase/models/schema"
)

func ptr[T any](v T) *T {
	return &v
}

func init() {
	log.Println("Registering initial schema migration...")
	m.Register(func(db dbx.Builder) error {
		dao := daos.New(db)

		// 1. Create announcements collection
		announcements := &models.Collection{}
		announcements.Name = "announcements"
		announcements.Type = models.CollectionTypeBase
		announcements.System = false
		announcements.Schema = schema.NewSchema(
			&schema.SchemaField{
				Name:     "title",
				Type:     schema.FieldTypeText,
				Required: true,
				Options: &schema.TextOptions{
					Min: ptr(1), Max: ptr(200),
				},
			},
			&schema.SchemaField{
				Name:     "slug",
				Type:     schema.FieldTypeText,
				Required: true,
				Options: &schema.TextOptions{
					Min: ptr(1), Max: ptr(200),
					Pattern: `^[a-z0-9]+(?:-[a-z0-9]+)*$`,
				},
			},
			&schema.SchemaField{
				Name:     "content",
				Type:     schema.FieldTypeEditor,
				Required: true,
				Options: &schema.EditorOptions{
					ConvertUrls: true,
				},
			},
			&schema.SchemaField{
				Name:     "published_date",
				Type:     schema.FieldTypeDate,
				Required: true,
			},
			&schema.SchemaField{
				Name:     "featured",
				Type:     schema.FieldTypeBool,
				Required: false,
			},
		)
		// Add indexes via the Indexes field (requires raw SQL)
		announcements.Indexes = []string{
			"CREATE UNIQUE INDEX idx_unique_slug ON announcements (slug)",
		}
		announcements.ListRule = ptr("")
		announcements.ViewRule = ptr("")
		
		if err := dao.SaveCollection(announcements); err != nil {
			return err
		}

		// 2. Create documents collection
		documents := &models.Collection{}
		documents.Name = "documents"
		documents.Type = models.CollectionTypeBase
		documents.System = false
		documents.Schema = schema.NewSchema(
			&schema.SchemaField{
				Name:     "title",
				Type:     schema.FieldTypeText,
				Required: true,
				Options: &schema.TextOptions{
					Min: ptr(1), Max: ptr(200),
				},
			},
			&schema.SchemaField{
				Name:     "category",
				Type:     schema.FieldTypeSelect,
				Required: true,
				Options: &schema.SelectOptions{
					MaxSelect: 1,
					Values: []string{
						"Meeting Minutes",
						"Bylaws",
						"Financial Reports",
						"Architectural Guidelines",
					},
				},
			},
			&schema.SchemaField{
				Name:     "year",
				Type:     schema.FieldTypeNumber,
				Required: true,
				Options: &schema.NumberOptions{
					Min: ptr(2000.0), Max: ptr(2100.0),
				},
			},
			&schema.SchemaField{
				Name:     "file",
				Type:     schema.FieldTypeFile,
				Required: true,
				Options: &schema.FileOptions{
					MaxSelect: 1,
					MaxSize:   10485760, // 10MB
					MimeTypes: []string{"application/pdf"},
				},
			},
			&schema.SchemaField{
				Name:     "description",
				Type:     schema.FieldTypeText,
				Required: false,
				Options: &schema.TextOptions{
					Max: ptr(500),
				},
			},
		)
		documents.Indexes = []string{
			"CREATE INDEX idx_category ON documents (category)",
			"CREATE INDEX idx_year ON documents (year)",
		}
		documents.ListRule = ptr("")
		documents.ViewRule = ptr("")

		if err := dao.SaveCollection(documents); err != nil {
			return err
		}

		return nil
	}, func(db dbx.Builder) error {
		dao := daos.New(db)

		announcements, _ := dao.FindCollectionByNameOrId("announcements")
		if announcements != nil {
			if err := dao.DeleteCollection(announcements); err != nil {
				return err
			}
		}

		documents, _ := dao.FindCollectionByNameOrId("documents")
		if documents != nil {
			if err := dao.DeleteCollection(documents); err != nil {
				return err
			}
		}

		return nil
	}, "1733842800_initial_schema.go")
}
